### DISCUSSION TITLE
Value Oriented Programming Needs Implicits?

### DISCUSSION THREAD

**Discussion by turbolent** [hylo-lang/hylo#731] [Q&A]
@matklad wrote a very interesting blog post about mutable value semantics: https://matklad.github.io/2023/05/02/implicits-for-mvs.html

I was wondering about the ergonomics around second-class references myself, so this is very timely.

What would an "idiomatic" implementation of a sort function in (a future version of) Val look like?

**Comment by dabrahams** [comment]
We think implicits might make some kinds of code more ergonomic, but I think "needs" is overstating the case.  You can do value-oriented programming in Swift no problem, and there are no implicits.  I'm out of time to write it out right now but an idiomatic implementation in Val would be similar to an idiomatic implementation in Swift, which is pretty straightforward.  A Quicksort will be more elegant in Val than what Swift people will mostly recommend you do because mutating slices in Swift is inefficient (choose a pivot, partition on it, quicksort both partitions as slices, swap the pivot into place).

**Reply by turbolent** [reply]
Thank you for the quick reply!

It would be nice to see idiomatic examples for operations on non-tree shaped types, which "traditionally" would use references in the object graph, but got converted so they do not, like the tree structure which has a parent "reference" (`Node` in the post), or even types which do "point" to some "global"/"shared" value through an identifier (like `Name` points to `NameTable` in the post).

I only skimmed all the functionality that's available in Val. The blog post proposes to capture the "parent"/"shared" structure which is needed for the operation (like `NameTable` is needed to e.g. sort `Name`), but doesn't the capture in the closure create an implicit reference? Is this OK because Val allows this as long as it can reason about the lifetime of the closure and can ensure it does not escape?

**Reply by kyouko-taiga** [reply]
A thorough answer to your question would probably deserve its own blog post but here are some observations that might help getting a glimpse of the big picture.

In strict MVS, one way to represent arbitrary graphs is to use adjacency lists or matrices. One can then implement all kinds of APIs on top of this representation to emulate the look-and-feel of traditional reference-based APIs.

_Note: some algorithms are actually better expressed on adjacency lists than on reference webs._

For example, one could implement a tree that allows you to get from child to parent and vice versa by encapsulating all nodes and providing an index-based API to query its contents:

```swift
public fun main() {
  var tree = Tree(root: T())
  let child = &tree.add_leaf(T(), to: tree.root())
  assert(tree.parent(of: child) == tree.root())
}
```

In this example `child` and `tree.root` are indices identifying a node in the tree.

This kind of approach gets you surprisingly far. Things only get a little messy when you need to express operations on multiple elements of different data structures (e.g., an operation on a tree that also accesses the contents of different graph). In those situations, argument lists tend to grow long, leading to undesirable APIs. That's specifically where I think implicits might be of big help.

Going back to adjacency lists, one may argue we've just substituted indices for pointers/references. But doing so we gained safety, since `Tree` can verify that an index is in bound, and local reasoning, since one cannot change a part of `Tree` without access to the whole.

You can even build temporary "views" of these data structures by encapsulating the whole into an object whose API focuses on a part. For example, we could create a `SubTree` that encapsulates an entire tree (or another subtree) together with the index of its own root. Once we're done with a view, we can destroy it to recover the whole.

```swift
public fun main() {
  var t = some_tree()
  let u = tree.subtree(rooted_at: some_index())
  // <- accessing t here is illegal
  print(u.count)
  &t = u.release()
  print(t.count)
}
```

In this example, we _moved_ (i.e., transferred ownership) an entire tree to a subtree view, used that view, and then reinstated the tree. No reference was involved in the whole process because all values were treated as linear objects. No (expensive) copy was involved either.

From here, one observation we can make is that in the absence of any mutation, it's fine to allow `u` to be accesses simultaneously with `t`. That's because in this particular example, we can guarantee that `u` doesn't outlive `t`. So it's fine to capture a reference to `t` (or parts thereof) to form `u`. After all, if neither can change throughout that projection, then for all intents and purposes, both `t` and `u` can be thought as independent values.

Rust captures that intuition in its borrowing mechanism. Val uses projections, which we argue are conceptually simpler because, for all intents and purposes, a projection of `T` is another instance of `T`. It is does not have a different type, so there's no need for wrappers, proxies, dereferencing, etc.

Once mutation enters the scene, we can still make projections but we must ensure uniqueness. So that means we can "capture" a part of a whole in a mutable view as long as we can statically prove that the view doesn't outlive the whole and that the whole doesn't get accessed throughout the lifetime of the view.

**Reply by matklad** [reply]
> Things only get a little messy when you need to express operations on multiple elements of different data structure

The second case where this becomes messy is polymorphism. If `xs` is an array of items, then `"\(xs)"` does the right thing. If `xs` is a list of _indexes_ of items, then `"\(xs)"` often is not what you want (as it prints "pointers", rather than the actual items). You can implement `"\(xs)"` "by hand", but you can't re-use array's custom string convertible implementation.

**Reply by dabrahams** [reply]
You can look at any of the algorithms in Boost Graph Library for examples of operations on non-tree-shaped types using value semantics.  In Val algorithms on a graph would be methods of the graph, and when vertices and edges are passed around, the value actually passed is the vertex or edge ID, but you can just think of it as the vertex or the edge.  The `self` of the graph is what gives you the context for following edges.  The DFS implementations on [this page](https://www.programiz.com/dsa/graph-dfs) (even the Python one) are good examples of how it looks.  Very straightforward and like the pseudocode.  Also very efficient.

> doesn't the capture in the closure create an implicit reference?

For Val, it depends on your point of view; if you want to view it as a reference, you just need to remember that the compiler ensures the Law of Exclusivity is obeyed.  Mutations can only be observed through a mutable reference, and there can only ever be one of those, to any given object, available at a time.  The other point of view is that the value is captured, either mutably or immutably, and if mutably (and not consumed/sunk), it is notionally moved back to whence it came when the closure's lifetime ends.  With the 2nd point of view, Val has no references.

**Reply by dabrahams** [reply]
> The second case where this becomes messy is polymorphism. If `xs` is an array of items, then `"\(xs)"` does the right thing. If `xs` is a list of _indexes_ of items, then `"\(xs)"` often is not what you want (as it prints "pointers", rather than the actual items). You can implement `"\(xs)"` "by hand", but you can't re-use array's custom string convertible implementation.

The 2nd and 3rd sentences are certainly true but AFAICT it has no relationship to polymorphism.  It is simply a consequence of programming with value semantics, with or without polymorphism: you don't get direct access to from object x to objects that are not parts of x.  Using value semantics will sometimes force you to represent problems/solutions differently than you otherwise might, which is no more than one should expect: all disciplines that have safety/local reasoning benefits create similar effects on the programming model.  Structured programming => no gotos.

**Reply by turbolent** [reply]
I hope I was not mistaken, my question wasn't trying to argue that it is a problem that solutions have to be implemented differently than in languages following other paradigms, I was merely just curious how idiomatic solutions would look like in the MVS paradigm. Both your and Dimi's answers helped with that, thank you, I have a bit of a better sense how to approach modelling solutions in the paradigm now.

**Reply by dabrahams** [reply]
@turbolent I think you misunderstood me.  The point of [this post](https://github.com/val-lang/val-lang.github.io/discussions/61#discussioncomment-5796013) was merely to say that the connection—drawn by @matklad—between polymorphism and these observations about the programming model seems to me like the result of a misunderstanding (which is likely to lead to further misunderstanding).

**Reply by matklad** [reply]
FWIW, I carefully read @dabrahams's comments, and I stand by what I said, and the specific wording I used. My wording is fairly compressed and can be misunderstood of course. A more unpacked version of what I think is here: https://github.com/matklad/matklad.github.io/issues/122#issuecomment-1533741655

**Reply by kyouko-taiga** [reply]
@matklad, I think your statement adds unnecessary qualification to the approaches that get harder to express with MVS.

The "_certain (type-class based) approaches to generic programming_" that you mention aren't approaches suitable to MVS before even considering generic programming. If you want to use MVS, then you cannot store references to parts that confer access to their whole, even if you're writing programs that only manipulate `Int`s and arrays thereof.

So I think a more accurate statement should be "_value semantics makes certain (type-class based) approaches harder_", to which @dabrahams  answered "_which is no more than one should expect: all disciplines that have safety/local reasoning benefits create similar effects on the programming model_".

Hope that clarifies a possible misunderstanding, or at the very least doesn't bring any more confusion.

**Reply by matklad** [reply]
Thanks @kyouko-taiga! This indeed seems like a specific misunderstanding which is consistent with the comments so far! Let me elaborate this angle then!

It is true that value semantics makes _everything_ a bit more complicated as you don't have references. That's an "obvious point" which isn't mentioned in the post explicitly. For non-generic code, this is not a big deal: you just add a `ctx: Context` parameter to various concrete functions and that's it. It _can_ get a bit tedious if you have different kinds of contexts or need nested lookups or both. Pulling this example from one value-oriented codebase of mine:

```rust
// Indexes:
store.names.lookup(store.fun[*callee].name)
// Direct references:
callee.name
```

But even in complex _concrete_ cases, this all is fairly straightforward, you just add parameters to function declarations and fix call-sites.

For _generic_ code, the situation is meaningfully more interesting. Even in simple generic cases, you can not "just" add a context parameter, because the generic signature is fixed upstream. You can work-around that by having two copies of generic API, one that works based on _type_-level traits, and one that takes _value_-level context, but having two copies of generic API means that the API is not actually maximally generic!

So, the two _distinct_ difficulties arising from the forgoing of references are:

* you need to add context parameters to concrete functions (tedious, but straightforward)
* you need to add context parameters to generic functions (requires some imagination, as signatures are fixed)

**Reply by kyouko-taiga** [reply]
I think I understand what you mean.

I suppose higher-order functions can help tremendously here. `sort(by:)` is a good example that can be derived for many other generic algorithms:

```swift
let x = callees.first(where: fun(c) { store.fun[c].name == "abc" })
let y = callees.filter(where: fun(c) { store.fun[c].name == "abc" })
let z = callees.map(fun(c) { store.fun[c].name.count() })
...
```

You'll probably have multiple versions of these algorithms, but only for convenience. There's one maximally generic signature, which is the one that takes a lambda. That's because `callees.contains(x)` is sugar for `callees.contains(where: fun(c) { c == x })`, regardless of the type of `callees`'s elements.

**Reply by kyouko-taiga** [reply]
The next challenge is mutation, because the maximally generic variant of some algorithms will also have to consider effects as parameters. We plan on using polymorphic effects to address this need, but this feature probably won't land before the first release of the language.

```swift
extension Collection {
  public fun for_each<E, a: access>(_ action: [E](a Element) -> Void) a { ... }
}
```

In this example, `for_each(_:)` is generic over the mutability of the elements that it passes to `action`. If the latter takes `let` parameters, `for_each` is instantiated as a `let` (i.e., non-mutating) method. If it takes `inout` parameters, then `for_each` is instantiated as a `inout` (i.e., mutating) method.

**Reply by matklad** [reply]
>I suppose higher-order functions can help tremendously here. sort(by:) is a good example that can be derived for many other generic algorithms:

Exactly! And then you start imagining  a `by` version for something like printing a `Collection<Collection<IAmAnIndex>>`, and end up writing an article about implicits :)

**Comment by muscar** [comment]
That's a great explanation @kyouko-taiga. The blog post had an interesting example with a name table with strings in it where the external API exposes int handles. The problem posed was to sort a vector of such handles but in the lexicographic order given by the elements they represent. How would that look in Val?

By the way, I find Val very interesting, great work :)

**Reply by kyouko-taiga** [reply]
> That's a great explanation @kyouko-taiga.

Thanks.

> The blog post had an interesting example with a name table with strings in it where the external API exposes int handles. The problem posed was to sort a vector of such handles but in the lexicographic order given by the elements they represent. How would that look in Val?

That's the situation I mentioned (was probably still editing when you answered; concurrency is hard). When an operation on a whole needs access to the contents of other data structures, APIs tend to get messier.

I think implicits are a very solid response to this problem. There are alternative approaches. The simplest is of course to use a lambda to sort the vector of handles:

```swift
let table: Array = ["durian", "apple", "mango"]

public fun main() {
  var s = name_table.indices()
  &s.sort(comparing_with: fun(a, b) { table[a].lexicographically_precedes(table[b]) })
}
```

Another approach would be to "bundle" a projection of the table together with each individual handle. That would be similar to iterator in C++, but guaranteed safe. Rust can do it out of the box if I'm not mistaken. It's a little trickier to express in Val because we can't store projections in containers (at least currently). You could use the unsafe API though:

```swift
extension Array {
  public type BundledIndex {
    let b: Pointer<Element>
    let i: Array.Index
    property element: Element { unsafe b.advanced_by(i).pointee }
  }
  public with_bundled_indices<T>(_ do action: (Array<BundledIndex>) -> T) -> T {
    unsafe with_buffer_pointer(fun(b) {
      action(self.indices.map(fun(i) { BundledIndex(b, i) ))
    })
  }
}
```

Note that since `Array.BundledIndex` is not `Sinkable` nor `Copyable`, it cannot escape the lambda we pass to `Array.with_bundled_indices`.

**Reply by turbolent** [reply]
Thank you for the great explanation and the illustrative examples 👍

**Reply by turbolent** [reply]
@kyouko-taiga 
> Note that since Array.BundledIndex is not Sinkable nor Copyable, it cannot escape the lambda we pass to Array.with_bundled_indices.

I looked at https://forums.swift.org/t/template-for-a-possible-future-object-model/59823#the-copyable-and-sinkable-traits-14 to understand `Sinkable` a bit more.

So is the call of `action` not a problem, because `action` is effectively `let`, so that is not considered a sink, but if the closure passed to `action` would then e.g. try to bind it (or return it?), that would be a sink and fail?

**Reply by kyouko-taiga** [reply]
Because `Array<BundledIndex>` is not `Sinkable`, `action` cannot return it or store it anywhere. Let's try to break those assumptions:

```swift
let table: Array = ["durian", "apple", "mango"]
public fun main() {
  // error: cannot return `x`; `Array<BundledIndex>` does not conform to `Sinkable`
  let esc0 = table.with_bundled_indices(fun(x) { x })

  // error: `Array<BundledIndex>` has no method `copy`
  let esc1 = table.with_bundled_indices(fun(x) { x.copy() })

  // error: cannot assign `esc2`; `Array<BundledIndex>` does not conform to `Sinkable`
  var esc2: Array<BundledIndex>
  table.with_bundled_indices(fun(x) { &esc2 = x })
}
```

_Note: I've been recently thinking about the interaction between `sink` and `Sinkable` and came to the conclusion that "sinking" an argument might not necessarily mean to move it, which is what the trait `Sinkable` actually represent. Perhaps we should revisit our name choices._

**Reply by tjammer** [reply]
Since C++ iterators were mentioned:
There is a related discussion on how such things can be represented in Val https://github.com/orgs/hylo-lang/discussions/742
At the end, implicit parameter (like Scala's implicits) are even mentioned as a possible solution.
